### PR TITLE
APP-670-Spike

### PR DESCRIPTION
**Pull request by kvnvelasco**
This change is [Reviewable](https://reviewable.io/reviews/mathspace/msquill/29)

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -18,7 +18,8 @@ BASE_SOURCES = \
   $(SRC_DIR)/controller.js \
   $(SRC_DIR)/publicapi.js \
   $(SRC_DIR)/services/*.util.js \
-  $(SRC_DIR)/services/*.js
+  $(SRC_DIR)/services/*.js \
+	$(SRC_DIR)/configuration/symbols/*.js
```
Comment (bholloway) [inline Makefile]: **nit** too much indent?

File: build/mathquill.css [hunk 2]
```diff
@@ -0,0 +1,519 @@
+/*
+ * MathQuill: http://mathquill.com
```
Comment (bholloway) [inline build/mathquill.css]: This file is new. Is it intended to be in the `build`?

File: build/mathquill.js [hunk 3]
```diff
@@ -2110,43 +2277,55 @@ Controller.open(function(_, super_) {
...
-      var ultimateRootAPI = ultimateRoot.controller.API;
-
-      var rootjQ = $(e.target).closest('.mq-root-block');
+      // finds the closest parent root block to get an instance of Controller.
+      // If none exists, fallback to ultimate root controller
+      var rootjQ = $(e.target).closest('.mq-root-block'); 
       var root = Node.byId[rootjQ.attr(mqBlockId) || ultimateRootjQ.attr(mqBlockId)];
-      var ctrlr = root.controller, cursor = ctrlr.cursor, blink = cursor.blink;
-      var textareaSpan = ctrlr.textareaSpan, textarea = ctrlr.textarea;
+      var ctrlr = root.controller,
```
Comment (bholloway) [inline build/mathquill.js]: **nit** better to keep this and minimise change

File: build/mathquill.js [hunk 4]
```diff
@@ -2110,43 +2277,55 @@ Controller.open(function(_, super_) {
...
+          // and you clicked inside the editable box
+          ctrlr.API.latex().indexOf("\\editable{") > -1) {
+          // or if the selected element contains latex matching '\editable{'
+          // ignore the event (no blinking cursor)
+          return;
+        }
+      } // -- MATHSPACE
 
       function mousemove(e) { target = $(e.target); }
+
```
Comment (bholloway) [inline build/mathquill.js]: **nit** better to omit this and minimise change

File: build/mathquill.js [hunk 5]
```diff
@@ -2182,16 +2364,27 @@ Controller.open(function(_) {
...
+  /**
+   * 
+   */
   _.seek = function(target, pageX, pageY) {
-    var cursor = this.notify('select').cursor;
+    var cursor = this.cursor;
+
+    // Notify all listeners bound to the Controller that 
+    // a selection has started
+    this.notify('select')
```
Comment (bholloway) [inline build/mathquill.js]: **caution** Is this change necessary?

File: build/mathquill.js [hunk 6]
```diff
@@ -2317,7 +2510,7 @@ Controller.open(function(_) {
   _.editablesTextareaEvents = function() {
     var ctrlr = this, root = ctrlr.root, cursor = ctrlr.cursor,
       textarea = ctrlr.textarea, textareaSpan = ctrlr.textareaSpan;
-
+
```
Comment (bholloway) [inline build/mathquill.js]: **nit** better to omit this and minimise change

File: jsmvc/math_components/latex_editor/latex_editor.js [hunk 7]
```diff
@@ -113,7 +113,7 @@ define([
                     this.mathFieldAPI = MathQuill.MathField(elm[0], {
                         // todo: Options duplicated in widgets/math_input.js
                         autoCommands: 'pi and or lim',
-                        autoCommandsMapping: {'and': 'andword', 'or': 'orword'},
+                        hoMapping: {'and': 'andword', 'or': 'orword'},
```
Comment (bholloway) [inline jsmvc/math_components/latex_editor/latex_editor.js]: **question** What does this name mean? It looks like a typo. Is it a necessary change.

File: src/services/textarea.js [hunk 8]
```diff
@@ -70,7 +70,7 @@ Controller.open(function(_) {
   _.editablesTextareaEvents = function() {
     var ctrlr = this, root = ctrlr.root, cursor = ctrlr.cursor,
       textarea = ctrlr.textarea, textareaSpan = ctrlr.textareaSpan;
-
+
```
Comment (bholloway) [inline src/services/textarea.js]: **nit** better to omit this and minimise change

File: src/services/mouse.js [hunk 9]
```diff
@@ -75,16 +90,27 @@ Controller.open(function(_) {
...
+  /**
+   * 
+   */
   _.seek = function(target, pageX, pageY) {
-    var cursor = this.notify('select').cursor;
+    var cursor = this.cursor;
+
+    // Notify all listeners bound to the Controller that 
+    // a selection has started
+    this.notify('select')
```
Comment (bholloway) [inline src/services/mouse.js]: Again I am not sure we should be making such changes

File: src/services/mouse.js [hunk 10]
```diff
@@ -3,43 +3,55 @@
...
-      var ultimateRootAPI = ultimateRoot.controller.API;
-
-      var rootjQ = $(e.target).closest('.mq-root-block');
+      // finds the closest parent root block to get an instance of Controller.
+      // If none exists, fallback to ultimate root controller
+      var rootjQ = $(e.target).closest('.mq-root-block'); 
       var root = Node.byId[rootjQ.attr(mqBlockId) || ultimateRootjQ.attr(mqBlockId)];
-      var ctrlr = root.controller, cursor = ctrlr.cursor, blink = cursor.blink;
-      var textareaSpan = ctrlr.textareaSpan, textarea = ctrlr.textarea;
+      var ctrlr = root.controller,
```
Comment (bholloway) [inline src/services/mouse.js]: **nit** better to keep this and minimise change

File: src/services/keystroke.js [hunk 11]
```diff
@@ -5,8 +5,18 @@
...
   _.keystroke = function(key, evt) {
+    if(this.keystrokeHandlers) {
+      for(var i = 0; i < this.keystrokeHandlers.length; i++)
+        this.keystrokeHandlers[i].call(this, key, evt);
+    }
     this.cursor.parent.keystroke(key, evt, this);
   };
+  _.registerKeystrokeHandler = function(fn) {
+    if(!this.keystrokeHandlers) {
+      this.keystrokeHandlers = []
```
Comment (bholloway) [inline src/services/keystroke.js]: **opinion** Do we know when the start of the lifecycle is? What is `this` and when is it created? I would be nicer to initialise `this. keystrokeHandlers` at construction and not have to test if it exists all the time.

File: src/intro.js [hunk 12]
```diff
@@ -113,3 +113,5 @@ function bind(cons /*, args... */) {
 function pray(message, cond) {
   if (!cond) throw new Error('prayer failed: '+message);
 }
+
```
Comment (bholloway) [inline src/intro.js]: **nit** too many blank lines

File: src/publicapi.js [hunk 13]
```diff
@@ -69,12 +69,21 @@ var AbstractMathQuill = P(function(_) {
   };
   _.config =
   MathQuill.config = function(opts) {
-    for (var opt in opts) if (opts.hasOwnProperty(opt)) {
+    for (var opt in opts) if (opts.hasOwnProperty(opt) && opts[opt] != null) {
```
Comment (bholloway) [inline src/publicapi.js]: **caution** Are you sure that falsey values should be filtered out here? I would suggest not _enhancing_ anything here.

File: src/publicapi.js [hunk 14]
```diff
@@ -69,12 +69,21 @@ var AbstractMathQuill = P(function(_) {
...
     }
+    
+    if (opts.commands) setTimeout(function() {
+      this.__controller.extendLatexGrammar(opts.commands)
+    }.bind(this));
+
     return this;
   };
+  _.injectGrammar = function(list) {
+
```
Comment (bholloway) [inline src/publicapi.js]: **question** why blank function?

File: src/publicapi.js [hunk 15]
```diff
@@ -69,12 +69,21 @@ var AbstractMathQuill = P(function(_) {
...
   _.config =
   MathQuill.config = function(opts) {
-    for (var opt in opts) if (opts.hasOwnProperty(opt)) {
+    for (var opt in opts) if (opts.hasOwnProperty(opt) && opts[opt] != null) {
       var optVal = opts[opt], processor = optionProcessors[opt];
       this.__options[opt] = (processor ? processor(optVal) : optVal);
+      // this can be used for dynamic updates but we need to call the correct subsystems 
     }
+    
+    if (opts.commands) setTimeout(function() {
```
Comment (bholloway) [inline src/publicapi.js]: **nit** Does this codebase omit braces on `if` statements? AFAIK it is universally considered bad form to omit curlys on any block.

File: build/mathquill.js [hunk 16]
```diff
@@ -1087,6 +1105,7 @@ var EditableField = MathQuill.EditableField = P(AbstractMathQuill, function(_) {
     return this;
   };
   _.typedText = function(text) {
+
```
Comment (bholloway) [inline build/mathquill.js]: **nit** better to omit this and minimise change

File: build/mathquill.js [hunk 17]
```diff
@@ -2182,16 +2364,27 @@ Controller.open(function(_) {
 });
 
 Controller.open(function(_) {
+  /**
```
Comment (bholloway) [inline build/mathquill.js]: **nit** write something here or remove comment

File: build/mathquill.js [hunk 18]
```diff
@@ -2834,16 +3203,28 @@ var MathBlock = P(MathElement, function(_, super_) {
...
+    
+    var latexCmds = controller.latexCmds
+    var charCmds = controller.charCmds
     var cons;
+    
     // exclude f because it gets a dedicated command with more spacing
-    if (ch.match(/^[a-eg-zA-Z]$/))
-      return Letter(ch);
+    if (ch.match(/^[a-eg-zA-Z\+\-\!\=]$/))
+      cons = Letter(ch);
```
Comment (bholloway) [inline build/mathquill.js]: **nit** Is this change necessary? You are converting this to mutable code.

File: build/mathquill.js [hunk 19]
```diff
@@ -2834,16 +3203,28 @@ var MathBlock = P(MathElement, function(_, super_) {
...
     else if (/^\d$/.test(ch))
-      return Digit(ch);
-    else if (cons = CharCmds[ch] || LatexCmds[ch])
-      return cons(ch);
+      cons = Digit(ch);
+    else if (cons = controller.searchForCommand(ch) )
```
Comment (bholloway) [inline build/mathquill.js]: **opinion** Assignment in an `if` is super nasty (pre-existing). Would be better to place some curlys on this block and do it properly. Meaning:
```
else {
  var fn = controller.searchForCommand(ch);
  return fn ? fn(ch) : VanillaSymbol(ch);
}
```

File: build/mathquill.js (hunk 3 above)
Comment (kvnvelasco) [reply build/mathquill.js]: This is the build tho.

File: src/services/mouse.js (hunk 9 above)
Comment (kvnvelasco) [reply src/services/mouse.js]: This was more for code navigation purposes. `this.notify` returns `this` when called (probably for chaining). The existing code was ambiguous combining a method call and an assignment into one line.

I think this is spill over from when I was still studying the code. It would be a shame to lose the signposts for anyone else in the future trying to learn this.

File: src/services/keystroke.js (hunk 11 above)
Comment (kvnvelasco) [reply src/services/keystroke.js]: The lifecycle here is a bit tricky and it would need some modification to the core controller code. We'd have to create a new method like `_.initKeystrokeHandlers` and the call that from when the controller is initialized

File: src/services/mouse.js (hunk 9 above)
Comment (bholloway) [reply src/services/mouse.js]: agreed

File: src/services/mouse.js [hunk 20]
```diff
@@ -3,43 +3,53 @@
  *******************************************************/
 
 Controller.open(function(_) {
+  /**
+   *
```
Comment (bholloway) [inline src/services/mouse.js]: **nit** if this comment is to remain empty I think it should be omitted

File: build/mathquill.js (hunk 19 above)
Comment (bholloway) [reply build/mathquill.js]: wrong file. I'm happy with the src file.

File: src/configuration/symbols/symbols.utils.js [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+        name: symbolDefinition.name,
+        latex: '',
+        htmlEntity: '',
+        match: symbolDefinition.match
+      }
+
+    var boundSymbol = binder(symbolDefinition);
+
+    if (symbolDefinition.match)
+      symbolDefinition.match.forEach(function (match) {
```
Comment (bholloway) [inline src/configuration/symbols/symbols.utils.js]: **opinion** Better as `symbolDefinition.match.some()`

File: src/configuration/symbols/symbols.utils.js [hunk 22]
```diff
@@ -0,0 +1,72 @@
...
+        if (existingSymbols[match])
+          console.warn("Symbol definition duplication in " + match);
+        symbols[match] = boundSymbol
+      })
+    
+    // In some cases we can opt not to have a name clause. This is used for match only 
+    // definitions. See: GREEK_SYMBOLS
+    if (symbolDefinition.name) {
```
Comment (bholloway) [inline src/configuration/symbols/symbols.utils.js]: **question** Does this case go away if we require `match` to be present?

File: src/publicapi.js [hunk 23]
```diff
@@ -69,10 +69,18 @@ var AbstractMathQuill = P(function(_) {
...
+    for (var opt in opts) if (opts.hasOwnProperty(opt) && opts[opt] != null) {
       var optVal = opts[opt], processor = optionProcessors[opt];
       this.__options[opt] = (processor ? processor(optVal) : optVal);
+      // this can be used for dynamic updates but we need to call the correct subsystems 
+    }
+    
+    if (opts.commands) {
+      setTimeout(function () {
+        this.__controller.extendLatexGrammar(opts.commands)
+      }.bind(this));
```
Comment (bholloway) [inline src/publicapi.js]: **opinion** Can you achieve this without the additional function?
```
setTimeout(this.__controller.extendLatexGrammar.bind(this, opts.commands), 0);
```

Comment (bholloway) [inline src/publicapi.js]: **opinion** optional args are evil, prefer explicit `0`

File: src/services/commands.js [hunk 24]
```diff
@@ -0,0 +1,153 @@
...
+       
+        if (pattern === grammarDefinition.keystroke) {
+          event.preventDefault();
+          this.API.cmd(grammarDefinition.name);
+        }
+      });
+      // we want to make sure that keystroke inputs do not generate any markup 
+      // this means overriding the 
+    }
+
```
Comment (bholloway) [inline src/services/commands.js]: **nit** Okay so maybe I have many nits about blank lines, but this is ridiculous! :wink:

File: src/services/keystroke.js [hunk 25]
```diff
@@ -5,8 +5,18 @@
 
 Controller.open(function(_) {
   _.keystroke = function(key, evt) {
+    if(this.keystrokeHandlers) {
```
Comment (bholloway) [inline src/services/keystroke.js]: **nit** missing space after `if`

File: src/services/keystroke.js [hunk 26]
```diff
@@ -5,8 +5,18 @@
...
+      for(var i = 0; i < this.keystrokeHandlers.length; i++)
```
Comment (bholloway) [inline src/services/keystroke.js]: **nit** missing space after `for`

File: src/services/keystroke.js [hunk 27]
```diff
@@ -5,8 +5,18 @@
...
+        this.keystrokeHandlers[i].call(this, key, evt);
+    }
     this.cursor.parent.keystroke(key, evt, this);
   };
+  _.registerKeystrokeHandler = function(fn) {
+    if(!this.keystrokeHandlers) {
```
Comment (bholloway) [inline src/services/keystroke.js]: **nit** missing space after `if`

File: src/services/commands.js [hunk 28]
```diff
@@ -0,0 +1,153 @@
...
+    this.textCommands = {}
+
+    this.cursor.options.autoCommands = {}
+    
+    // Initialize the grammar processors for various symbols
+    // by loading up the default configuration 
+    // This is WIP. Eventually all files symbol definitions
+    // should be transferred over to the defaultSymbolDefinitions file
+    this.extendLatexGrammar(GLOBALLY_DISABLED_INPUT, 'symbol');
+    // this.extendLatexGrammar(VANILLA_SYMBOLS, "symbol");
```
Comment (bholloway) [inline src/services/commands.js]: Does this need to be completed as part of this release?  If not we will need to track it with a Jira issue.

File: src/services/commands.js [hunk 29]
```diff
@@ -0,0 +1,153 @@
...
+  binary: symbolFactory(function(symboldefinition) {
+    return bind(
+      BinaryOperator,
+      symboldefinition.latex,
+      symboldefinition.htmlEntity
+    );
+  })
+};
+
+
```
Comment (bholloway) [inline src/services/commands.js]: **nit** too many trailing blank lines

**Comment by bholloway** [comment]
Needs some basic linting on the changed files. At least semicolons.

Maybe an oldskool jshint cli or something that doesn't need much configuration and isn't too opinionated.

**Comment by bholloway** [comment]
I really have no idea what `reviewable` is doing. It is showing me changes from all files yet some changes are not there.

All we can do is try to converge on this.
